server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,478 @@
+                    if (verbose()) {
+                        message(String.format(Locale.ROOT, "merge task %s start", this));
+                    }
+                    try {
+                        doMerge(mergeSource, onGoingMerge.getMerge());
+                        if (verbose()) {
+                            message(
+                                String.format(
+                                    Locale.ROOT,
+                                    "merge task %s merge segment [%s] done estSize=%.1f MB (written=%.1f MB) "
+                                        + "runTime=%.1fs (stopped=%.1fs, paused=%.1fs) rate=%s",
+                                    this,
+                                    getSegmentName(onGoingMerge.getMerge()),
+                                    bytesToMB(onGoingMerge.getMerge().estimatedMergeBytes),
+                                    bytesToMB(rateLimiter.getTotalBytesWritten()),
+                                    nsToSec(System.nanoTime() - mergeStartTimeNS.get()),
+                                    nsToSec(rateLimiter.getTotalStoppedNS()),
+                                    nsToSec(rateLimiter.getTotalPausedNS()),
+                                    rateToString(rateLimiter.getMBPerSec())
+                                )
+                            );
+                        }
+                    } finally {
+                        long tookMS = TimeValue.nsecToMSec(System.nanoTime() - mergeStartTimeNS.get());
+                        mergeTracking.mergeFinished(onGoingMerge.getMerge(), onGoingMerge, tookMS);
+                    }
+                } finally {
+                    if (verbose()) {
+                        message(String.format(Locale.ROOT, "merge task %s end", this));
+                    }
+                    afterMerge(onGoingMerge);
+                }
+            } finally {
+                if (verbose()) {
+                    message(String.format(Locale.ROOT, "merge task %s end", this));
+                }
+                mergeDone(this);
+                try {
+                    // kick-off next merge, if any
+                    merge(mergeSource, MergeTrigger.MERGE_FINISHED);
+                } catch (@SuppressWarnings("unused") AlreadyClosedException ace) {
+                    // OK, this is what the {@code ConcurrentMergeScheduler} does
+                }
+            }
+        }
+
+        void abortOnGoingMerge() {
+            // This interrupts the IndexWriter.
+            onGoingMerge.getMerge().setAborted();
+            // This ensures {@code OneMerge#close} gets invoked.
+            // {@code IndexWriter} considers a merge as "running" once it has been pulled from the {@code MergeSource#getNextMerge},
+            // so in theory it's not enough to just call {@code MergeSource#onMergeFinished} on it (as for "pending" ones).
+            doMerge(mergeSource, onGoingMerge.getMerge());
```

Comment:
@henningandersen I think this is the "theoretically" right implementation to abort a merge since `IndexWriter#abortOneMerge` is not exposed. Note that this is different from the threadpool-based merge scheduler implementation in serverless https://github.com/elastic/elasticsearch-serverless/blob/998e827b3ac09ca7b237d40b9b6b464967d86225/modules/stateless/src/main/java/co/elastic/elasticsearch/stateless/engine/ThreadPoolMergeScheduler.java#L243 . In practice, though, with the current `IndexWriter` implementation there should be no difference between the two, because the `IndexWriter` doesn't open segment readers when the `MergeSource#getNextMerge` is invoked.

Reply:
Seems reasonable to me, are you suggesting to maybe update the serverless one to do the same? (not saying we should now, I would sort of hope we get to just one scheduler out of this work eventually).

Reply:
> Seems reasonable to me, are you suggesting to maybe update the serverless one to do the same?

Yes, I think this is the safest, no-assumption approach. The merge code in the `IndexWriter` already has to deal with aborting merges at any point-in-time, including before they're even started.

> I would sort of hope we get to just one scheduler out of this work eventually)

++